### PR TITLE
A pre-commit hook to magically fix whitespace issues.

### PR DESCRIPTION
**Pull request by SkySkimmer**
It's supposed to leave non git-add'ed changes alone (using `git stash` while working).
It detects whether there is a merge, cherry-pick or rebase in progress. If interactive rebase it can work around it, otherwise it prints any whitespace error (`git diff --check`) and accepts the commit as is.

There may be a smarter way to do this, improvements welcome. I'm going to try it out for some time.

### CODE REVIEW

**Comment by SkySkimmer** [comment]
>I'm going to try it out for some time.

Note however that emacs already eats my whitespace errors so this isn't the best testing in the world.

**Comment by psteckler** [comment]
The man page for `git diff` says that whitespace errors issued for the `--check` flag are controlled by the `core.whitespace` configuration. So maybe there should be such a config that works with this hook, so all users see the same errors.

**Comment by SkySkimmer** [comment]
.gitattributes should be doing that.

**Comment by psteckler** [comment]
> .gitattributes should be doing that.

Right, I'm suggesting that the `core.whitespace` entry for `.gitattributes` could be added to this PR.

**Comment by Zimmi48** [comment]
@SkySkimmer It would probably make sense to suggest using this commit hook in the Contributing guide, wouldn't it?

**Comment by SkySkimmer** [comment]
Actually I think using git apply instead of rebase would be more robust so this isn't finished.

**Comment by SkySkimmer** [comment]
New version based on `git apply`.
`git apply --cached` is very nice and lets us cleanly keep staged and unstaged changes separate.

**Comment by Zimmi48** [comment]
Is this now ready and sufficiently tested? It would probably be good if this could be merged and #6268 be closed.

**Comment by SkySkimmer** [comment]
It's stable unless someone asks for changes. As I said my editor reduces test coverage but it hasn't broken my commits yet. I don't know if anyone else has been trying it out.
The `needs: review` is also still relevant.

**Comment by SkySkimmer** [comment]
In #6528 it was suggested to make configure install the hook. I'm not working on that so feel free to push a commit doing it.

**Comment by JasonGross** [comment]
I made https://github.com/SkySkimmer/coq/pull/5 to have `./configure` add the precommit hook automatically.

File: dev/tools/pre-commit [hunk 1]
```diff
@@ -0,0 +1,38 @@
+#!/bin/sh
+
+# Copy to .git/hooks/ to use.
+
+set -e
+
+# NB: even though we use --cached it seems to exit code 1 with
+# unstaged whitespace errors. That just means they get fixed though.
+if git diff-index --check --cached HEAD --quiet;
```
Comment (JasonGross) [inline dev/tools/pre-commit]: I think this is the wrong check.  `--quiet` implies `--exit-code`, which means that this command fails when there are any changes whatsoever.  So this is basically just a check that the commit is not empty.  The whitespace fixer is fast enough that this doesn't really hurt, though.

Comment (SkySkimmer) [reply dev/tools/pre-commit]: >I think this is the wrong check. --quiet implies --exit-code, which means that this command fails when there are any changes whatsoever. So this is basically just a check that the commit is not empty.

I guess that explains what I saw in the comment.

**Comment by SkySkimmer** [comment]
Does anyone know if commit hooks work without issues on windows?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -30,6 +30,8 @@ It's helpful to run the Coq test suite with `make test-suite` before submitting
 
 Don't be alarmed if the pull request process takes some time. It can take a few days to get feedback, approval on the final changes, and then a merge. Coq doesn't release new versions very frequently so it can take a few months for your change to land in a released version. That said, you can start using the latest Coq `master` branch to take advantage of all the new features, improvements, and fixes.
 
+Whitespace discipline (do not indent using tabs, no trailing spaces) is checked by Travis (using `git diff --check`). You can ensure that your commits do not introduce errors by adding the [`dev/tools/pre-commit`](/dev/tools/pre-commit) script to your `.git/hooks/` directory.
+
```
Comment (Zimmi48) [inline CONTRIBUTING.md]: Maybe add "just running configure" will add this, if you do not have a pre-existing pre-commit hook.

File: dev/tools/check-eof-newline.sh [hunk 3]
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env bash
 
-if [ -z "$(tail -c 1 "$1")" ]
+if ! dev/tools/should-check-whitespace.sh "$1" || [ -z "$(tail -c 1 "$1")" ]
```
Comment (Zimmi48) [inline dev/tools/check-eof-newline.sh]: Didn't you remove this file in another PR of yours?

Comment (SkySkimmer) [reply dev/tools/check-eof-newline.sh]: >Didn't you remove this file in another PR of yours?

#6568 which isn't merged yet.
This PR and that one conflict so I'll just fix whichever gets merged second.

**Comment by SkySkimmer** [comment]
Rebased.
I merged the `fix-eof-newline` and `check-eof-newline` by adding a `--fix` argument to the later.

File: dev/tools/pre-commit [hunk 4]
```diff
@@ -0,0 +1,47 @@
+#!/bin/sh
+
+# Copy to .git/hooks/ to use.
```
Comment (Zimmi48) [inline dev/tools/pre-commit]: Maybe update this comment.

File: dev/tools/pre-commit [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+set -e
+
+CODE=0
+git diff --cached --name-only -z | xargs -0 dev/tools/check-eof-newline.sh --fix || CODE=1
+
+if [ $CODE -ne 0 ]
+then
+    2>&1 echo "Some files had newline errors; they have been fixed in the working tree."
+    2>&1 echo "Make sure to add them before committing."
+    2>&1 echo "This may fix itself if you were using git commit -a, and you try again."
```
Comment (Zimmi48) [inline dev/tools/pre-commit]: Not fully clear what is meant here.

File: dev/tools/pre-commit [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+    git diff-index -p HEAD > "$TEMP"
+    git apply -R "$TEMP"
+    git apply --whitespace=fix "$TEMP"
+    rm "$TEMP"
+
+    # Check that we did fix whitespace
+    if git diff-index --check --cached HEAD;
+    then
+        :
```
Comment (Zimmi48) [inline dev/tools/pre-commit]: Does all POSIX shell have `!`? In that case, you could also just invert the logic of the conditional.

**Review by Zimmi48 — COMMENTED** [review]
General LGTM although I don't fully understand all the `git` magic. I haven't tested the last version but I bet that you have.

File: dev/tools/pre-commit (hunk 4 above)
Comment (SkySkimmer) [reply dev/tools/pre-commit]: >Maybe update this comment.

Done.

File: dev/tools/pre-commit (hunk 5 above)
Comment (SkySkimmer) [reply dev/tools/pre-commit]: >Not fully clear what is meant here.

The errors were fixed only in the worktree, so there was a missing `git add`.
Some thought had to be used to do that `git add` without adding other unstaged changes.
This is now fixed.

File: dev/tools/pre-commit (hunk 6 above)
Comment (SkySkimmer) [reply dev/tools/pre-commit]: >Does all POSIX shell have `!`? In that case, you could also just invert the logic of the conditional.

Done.

**Comment by SkySkimmer** [comment]
Someone check on that last commit, then we should be good to go.

**Comment by mattam82** [comment]
When trying to commit something which needs ws-fixing I do: stage, commit in magit, write a log message and then I get:
  1 /usr/local/bin/git … commit --
Auto fixing whitespace issues...
/var/folders/v4/4jfq8t4x4jnb_fs99j19v3300000gn/T/tmp.R92fG2Dv:10: trailing whitespace.

error: kernel/univ.ml: does not match index

That makes some sense because indeed the file has changed. The worst part is that the state after this has lost my changes. I would suggest making the script resistant to these kind of failures and save the users work somewhere first.

**Comment by SkySkimmer** [comment]
Technically changes aren't lost, they're in `tmp.R92fG2Dv` (autogenerated name).

Working on it.

**Comment by SkySkimmer** [comment]
OK I somehow completely misunderstood what `apply --index` did. It should be fixed now.

**Comment by mattam82** [comment]
If you make a patch that just introduces whitespace at the end of a line:
Auto fixing whitespace issues...
Patches are saved in 'git-fix-ws-index.VSg2r', 'git-fix-ws-index-fixed.Yu3Bp' and 'git-fix-ws-tree.XEcNj'.
If an error destroyes your changes you can recover using them.
(The files are cleaned up on success.)
git-fix-ws-index.VSg2r:10: trailing whitespace.
let compare_levels = Level.compare  
warning: 1 line applied after fixing whitespace errors.
git-fix-ws-index.VSg2r:10: trailing whitespace.
let compare_levels = Level.compare  
warning: 1 line applied after fixing whitespace errors.
error: unrecognized input

**Comment by mattam82** [comment]
The files are there with the (useless in this case) data. Otherwise no problem found by testing this a bit.

**Comment by mattam82** [comment]
Destroyed or destroys BTW

**Comment by SkySkimmer** [comment]
>Destroyed or destroys BTW

Fixed, also added commits:
- pre-commit: fail gracefully if fixing whitespace removes all changes
- pre-commit: nicer messages

**Comment by Zimmi48** [comment]
Where are we with this pre-commit hook? It would be great to merge because the linter continues to bother people and if we don't fix this soon, they will again try to rebel to have the linter removed (cf. https://github.com/coq/wiki/commit/4fecbe1712284101e745b97f80fdf4c92baf044b).

**Comment by SkySkimmer** [comment]
needs: review for 8cefdd3

**Comment by Zimmi48** [comment]
This particular commit LGTM. If you want some more testing, I can do that.

**Comment by SkySkimmer** [comment]
OK then.

**Pull request merged**